src/code/speed_meter.c

```diff
@@ -1,18 +1,277 @@
+/**
+ * How much time the RDP ran for over the course of `gGraphUpdatePeriod`.
+ */
+volatile OSTime gRDPTimeTotal;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/speed_meter/SpeedMeter_DrawTimeEntries.s")
+/**
+ * How much time elapsed between the last two `Graph_Update` ending.
+ * This is expected to be at least the duration of a single frame, since it includes the time spent waiting on the
+ * graphics task to be done.
+ */
+volatile OSTime gGraphUpdatePeriod;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/speed_meter/func_80177A84.s")
+// Accumulator for `gRSPAudioTimeTotal`
+volatile OSTime gRSPAudioTimeAcc;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/speed_meter/func_80177AC8.s")
+// Accumulator for `gRSPGfxTimeTotal`.
+volatile OSTime gRSPGfxTimeAcc;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/speed_meter/SpeedMeter_DrawAllocEntries.s")
+volatile OSTime gRSPOtherTimeAcc;
+volatile OSTime D_801FBB18;
+
+// Accumulator for `gRDPTimeTotal`
+volatile OSTime gRDPTimeAcc;
+
+typedef struct {
+    /* 0x0 */ volatile OSTime* time;
+    /* 0x4 */ u16 x;
+    /* 0x5 */ u16 y;
+    /* 0x6 */ u16 color;
+} SpeedMeterTimeEntry; // size = 0x8
+
+SpeedMeterTimeEntry* sSpeedMeterTimeEntryPtr;
+
+SpeedMeterTimeEntry sSpeedMeterTimeEntryArray[] = {
+    { &gRSPAudioTimeTotal, 0, 6, GPACK_RGBA5551(0, 0, 255, 1) },
+    { &gRSPGfxTimeTotal, 0, 8, GPACK_RGBA5551(255, 128, 128, 1) },
+    { &gRDPTimeTotal, 0, 10, GPACK_RGBA5551(0, 255, 0, 1) },
+    { &gGraphUpdatePeriod, 0, 12, GPACK_RGBA5551(255, 0, 255, 1) },
+};
+
+typedef struct {
+    /* 0x00 */ s32 maxval;
+    /* 0x04 */ s32 val;
+    /* 0x08 */ u16 backColor;
+    /* 0x0A */ u16 foreColor;
+    /* 0x0C */ s32 ulx;
+    /* 0x10 */ s32 lrx;
+    /* 0x14 */ s32 uly;
+    /* 0x18 */ s32 lry;
+} SpeedMeterAllocEntry; // size = 0x1C
+
+//! FAKE: if(1) in macro
+
+#define gDrawRect(gfx, color, ulx, uly, lrx, lry)                      \
+    do {                                                               \
+        if (gSysCfbHiResEnabled == true) {                             \
+            u32 tmp = color;                                           \
+            gDPPipeSync(gfx);                                          \
+            gDPSetFillColor(gfx, ((tmp) << 16) | (tmp));               \
+            gDPFillRectangle(gfx, (ulx)*2, (uly)*2, (lrx)*2, (lry)*2); \
+            if (1) {}                                                  \
+        } else {                                                       \
+            u32 tmp = color;                                           \
+            gDPPipeSync(gfx);                                          \
+            gDPSetFillColor(gfx, ((tmp) << 16) | (tmp));               \
+            gDPFillRectangle(gfx, (ulx), (uly), (lrx), (lry));         \
+        }                                                              \
+    } while (0)
+
+#define SET_FULLSCREEN_VIEWPORT_ALT(view)  \
```

Comment:
These two alt macros are a bit annoying :( should this one go in z64view.h?

Reply:
Hmm, now that I think about it yeah. `SET_FULLSCREEN_VIEWPORT_ALT` is actually meant to be used for dynamic fullscreen (i.e. it can support both hires and lores). Tbh, I kind of want to name this without the `ALT` and change the other one to `SET_FULLSCREEN_VIEWPORT_LORES` since it only support lores. Thoughts?

Reply:
The other option is name `SET_FULLSCREEN_VIEWPORT_ALT` `SET_FULLSCREEN_VIEWPORT_HIRES` since it supports hires.

Reply:
So the main difference between `SET_FULLSCREEN_VIEWPORT_HIRES` and `SET_FULLSCREEN_VIEWPORT` is `gScreenHeight` vs `SCREEN_HEIGHT`? Could `gScreenHeight` be either lores or hires? I'm not sure how `gScreenHeight` actually works in practice (and in parallel `gScreenWidth`)

Reply:
Correct that is the only difference. When LoRes or HiRes gets set through sys_cfb, both `gScreenHeight` and `gScreenWidth` get updated to the appropriate values (see `SysCfb_SetLoResMode` and `SysCfb_SetHiResMode`.

Reply:
I'm just thinking, because `SET_FULLSCREEN_VIEWPORT_HIRES` could be either HiRes or LowRes right? It's not just limited to HiRes as the name implies?

Reply:
Yeah hence we could go with the first option (rename the other one to `SET_FULLSCREEN_VIEWPORT_LORES`). I chose the second option mainly because the `SET_FULLSCREEN_VIEWPORT_LORES` is used so much since the game more often than not doesn't care about lores vs hires (only bombers notebook I think), and it matches OoT better. If you prefer I can go with that option though.

Reply:
Ah sorry that's not what I meant. The other one can be left alone. I mean you can use `SET_FULLSCREEN_VIEWPORT_HIRES` and it would set it to LoRes, which seems misleading.